### PR TITLE
[14.0][ADD] mail_send_error_activity

### PR DESCRIPTION
**Pull request by geomer198**

### CODE REVIEW

File: mail_send_error_activity/models/res_config_settings.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+from odoo import fields, models
+
+
+class ResConfigSettings(models.TransientModel):
+    _inherit = "res.config.settings"
+
+    activity_on_mail_error = fields.Boolean(
+        string="Activity on Error",
+        config_parameter="mail_send_error_activity.activity_on_mail_error",
```
Comment (ivs-cetmix) [inline mail_send_error_activity/models/res_config_settings.py:9]: формат параметра должен быть "module_name.field_name"

File: mail_send_error_activity/tests/test_mail_mail.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+            }
+        )
+
+        def unlink_replacement(self):
+            return
+
+        self.env["mail.mail"]._patch_method("unlink", unlink_replacement)
+
+    def test_create_activity_on_mail_error_invalid(self):
+        self.env["ir.config_parameter"].sudo().set_param(
```
Comment (isserver1) [inline mail_send_error_activity/tests/test_mail_mail.py:63]: Что делает данный тест?

**Review by ivs-cetmix — CHANGES_REQUESTED** [review]
См комментарии

File: mail_send_error_activity/models/mail_mail.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+    def _prepare_mail_error_activity_vals(self):
+        """
+        Prepare mail error Activity Vals
+        :return: dict
+        """
+        vals = dict()
+        if not (self.res_id and self.model):
+            return vals
+        # Model has attribute 'attribute_ids'
```
Comment (isserver1) [inline mail_send_error_activity/models/mail_mail.py:15]: Model has attribute 'activity_ids'

File: mail_send_error_activity/models/mail_mail.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        if not hasattr(self.env[self.model], "activity_ids"):
+            return vals
+        activity_type = self.env.ref(
+            "mail_send_error_activity.mail_activity_data_mail_send_error"
+        )
+        vals.update(
+            {
+                "user_id": self.author_id.id,
```
Comment (isserver1) [inline mail_send_error_activity/models/mail_mail.py:23]: Какая модель в поле user_id и какая в self.author_id?

File: mail_send_error_activity/models/mail_mail.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+                "note": self.failure_reason,
+                "res_model_id": self.env["ir.model"]._get(self.model).id,
+                "res_id": self.res_id,
+            }
+        )
+        return vals
+
+    def _create_activity_on_mail_error(self):
+        """
+        Create Activity Mail send error
```
Comment (isserver1) [inline mail_send_error_activity/models/mail_mail.py:38]: Create activity on mail send error

File: mail_send_error_activity/models/mail_mail.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        :return: bool
+        """
+        # Prepare activity value
```
Comment (isserver1) [inline mail_send_error_activity/models/mail_mail.py:41]: values

File: mail_send_error_activity/models/mail_mail.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        vals = self._prepare_mail_error_activity_vals()
+        if not vals:
+            return False
+        # Create activity Mail send error
```
Comment (isserver1) [inline mail_send_error_activity/models/mail_mail.py:45]: Create activity

**Review by isserver1 — CHANGES_REQUESTED** [review]

File: mail_send_error_activity/models/mail_mail.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+class MailMail(models.Model):
+    _inherit = "mail.mail"
+
+    def _prepare_mail_error_activity_vals(self):
+        """
+        Prepare mail error activity values
+        :return: dict of values
+        """
+        user_ids = self.author_id.user_ids
```
Comment (GabbasovDinar) [inline mail_send_error_activity/models/mail_mail.py]: it's not ids it's recordset, we need use "users" variable name

File: mail_send_error_activity/models/mail_mail.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        if (
+            len(user_ids) == 0
```
Comment (GabbasovDinar) [inline mail_send_error_activity/models/mail_mail.py]: why we need check length? Just "if not users"

File: mail_send_error_activity/models/mail_mail.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+            or not (self.res_id and self.model)
+            or not hasattr(self.env[self.model], "activity_ids")
```
Comment (GabbasovDinar) [inline mail_send_error_activity/models/mail_mail.py]: Can we use subclass?
```python
not issubclass(self.pool[model], self.pool['mail.activity.mixin'])
```

File: mail_send_error_activity/models/mail_mail.py [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+            return False
+        activity_type = self.env.ref(
+            "mail_send_error_activity.mail_activity_data_mail_send_error"
+        )
+        return {
+            "user_id": user_ids[0].id,
+            "activity_type_id": activity_type.id,
+            "summary": "{} {}".format(
+                activity_type.summary,
+                "".join([partner.email for partner in self.partner_ids]),
```
Comment (GabbasovDinar) [inline mail_send_error_activity/models/mail_mail.py]: maybe better join as ```", ".join(...```

File: mail_send_error_activity/models/mail_mail.py [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+
+
+class MailMail(models.Model):
+    _inherit = "mail.mail"
+
+    def _prepare_mail_error_activity_vals(self):
+        """
+        Prepare mail error activity values
+        :return: dict of values
+        """
```
Comment (GabbasovDinar) [inline mail_send_error_activity/models/mail_mail.py:13]: self.ensure_one()

File: mail_send_error_activity/models/mail_mail.py [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+                "".join([partner.email for partner in self.partner_ids]),
+            ),
+            "note": self.failure_reason,
+            "res_model_id": self.env["ir.model"]._get(self.model).id,
+            "res_id": self.res_id,
+        }
+
+    def _create_activity_on_mail_error(self):
+        """
+        Create an activity on mail send error
```
Comment (GabbasovDinar) [inline mail_send_error_activity/models/mail_mail.py]: Why is it necessary to put in a separate method of creating activiti?

File: mail_send_error_activity/models/mail_mail.py [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+    ):
+        # Proceed on mail error only
+        # and if enabled in settings
+        if failure_type and self.env["ir.config_parameter"].sudo().get_param(
+            "mail_send_error_activity.activity_on_mail_error", False
+        ):
+            for rec in self.filtered(
+                lambda mail: mail.author_id and mail.state == "exception"
+            ):
+                rec._create_activity_on_mail_error()
```
Comment (GabbasovDinar) [inline mail_send_error_activity/models/mail_mail.py]: we can call `_prepare_mail_error_activity_vals` method and the call `create`

File: mail_send_error_activity/tests/test_mail_mail.py [hunk 15]
```diff
@@ -0,0 +1,200 @@
+from odoo.tests import TransactionCase, tagged
+
+
+@tagged("post_install", "-at_install")
+class TestMailMail(TransactionCase):
+    def setUp(self):
+        super(TestMailMail, self).setUp()
```
Comment (GabbasovDinar) [inline mail_send_error_activity/tests/test_mail_mail.py:7]: Please, disable tracking

```python
self.env = self.env(
            context=dict(
                self.env.context,
                tracking_disable=True,
                no_reset_password=True,
            )
        )
```

File: mail_send_error_activity/tests/test_mail_mail.py [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+                "body_html": "Test Mail Message Body",
+                "res_id": self.res_partner_target.id,
+                "model": self.res_partner_bob._name,
+            }
+        )
+
+        def unlink_replacement(self):
+            return
+
+        self.env["mail.mail"]._patch_method("unlink", unlink_replacement)
```
Comment (GabbasovDinar) [inline mail_send_error_activity/tests/test_mail_mail.py]:
```
self.env["mail.mail"]._patch_method(
     "unlink", lambda *args: True
)
```

**Review by GabbasovDinar — CHANGES_REQUESTED** [review]

File: mail_send_error_activity/models/mail_mail.py (hunk 8 above)
Comment (geomer198) [reply mail_send_error_activity/models/mail_mail.py]: Resolved

File: mail_send_error_activity/models/mail_mail.py (hunk 9 above)
Comment (geomer198) [reply mail_send_error_activity/models/mail_mail.py]: Resolved

File: mail_send_error_activity/models/mail_mail.py (hunk 12 above)
Comment (geomer198) [reply mail_send_error_activity/models/mail_mail.py:13]: Resolved

File: mail_send_error_activity/models/mail_mail.py (hunk 14 above)
Comment (geomer198) [reply mail_send_error_activity/models/mail_mail.py]: Resolved

File: mail_send_error_activity/tests/test_mail_mail.py (hunk 16 above)
Comment (geomer198) [reply mail_send_error_activity/tests/test_mail_mail.py]: Resolved

File: mail_send_error_activity/tests/test_mail_mail.py (hunk 15 above)
Comment (geomer198) [reply mail_send_error_activity/tests/test_mail_mail.py:7]: Resolved

File: mail_send_error_activity/models/mail_mail.py (hunk 13 above)
Comment (geomer198) [reply mail_send_error_activity/models/mail_mail.py]: Resolved

File: mail_send_error_activity/models/mail_mail.py (hunk 11 above)
Comment (geomer198) [reply mail_send_error_activity/models/mail_mail.py]: Resolved

File: mail_send_error_activity/models/mail_mail.py (hunk 10 above)
Comment (geomer198) [reply mail_send_error_activity/models/mail_mail.py]: Resolved

**Review by GabbasovDinar — APPROVED** [review]

**Review by ivs-cetmix — APPROVED** [review]

**Pull request closed without merging**